File: app/domain/entities/quality_assessment.py
"""Quality assessment entity."""

from dataclasses import dataclass
from typing import Any, Dict, Optional


@dataclass(frozen=True)
class QualityAssessment:
    """Result of image quality assessment.

    This is an immutable value object representing quality metrics.
    Following Single Responsibility Principle - only contains quality data.

    Attributes:
        score: Overall quality score (0-100), higher is better
        blur_score: Blur detection score (Laplacian variance)
        lighting_score: Lighting quality score (mean brightness)
        face_size: Face size in pixels (minimum dimension)
        is_acceptable: Whether quality meets minimum threshold
        yaw: Estimated head yaw angle in degrees (None if landmarks unavailable)
        pitch: Estimated head pitch angle in degrees (None if landmarks unavailable)
        pose_acceptable: Whether head pose is within acceptable range (None if unavailable)

    Quality Guidelines:
        - score 0-40: Poor (reject)
        - score 41-70: Fair (warn user)
        - score 71-100: Good (accept)

    Note:
        This class is immutable (frozen) to ensure data integrity.
    """

    score: float
    blur_score: float
    lighting_score: float
    face_size: int
    is_acceptable: bool
    yaw: Optional[float] = None
    pitch: Optional[float] = None
    pose_acceptable: Optional[bool] = None

    def __post_init__(self) -> None:
        """Validate quality assessment data."""
        if not 0 <= self.score <= 100:
            raise ValueError(f"Score must be 0-100, got {self.score}")

        if self.blur_score < 0:
            raise ValueError(f"Blur score cannot be negative: {self.blur_score}")

        if self.face_size < 0:
            raise ValueError(f"Face size cannot be negative: {self.face_size}")

    def get_quality_level(self) -> str:
        """Get quality level as string.

        Returns:
            Quality level: "poor", "fair", or "good"
        """
        if self.score < 40:
            return "poor"
        elif self.score < 71:
            return "fair"
        else:
            return "good"

    def get_issues(self, blur_threshold: float = 15.0, min_face_size: int = 60) -> Dict[str, Any]:
        """Get list of quality issues with actual thresholds.

        Args:
            blur_threshold: Actual blur threshold used by the assessor
            min_face_size: Actual minimum face size used by the assessor

        Returns:
            Dictionary of quality issues with details
        """
        issues = {}

        if self.blur_score < blur_threshold:
            issues["blur"] = {
                "description": "Image is too blurry",
                "score": self.blur_score,
                "threshold": blur_threshold,
            }

        if self.face_size < min_face_size:
            issues["face_size"] = {
                "description": "Face is too small",
                "size": self.face_size,
                "minimum": min_face_size,
            }

        if self.lighting_score < 50:
            issues["lighting"] = {
                "description": "Poor lighting conditions",
                "score": self.lighting_score,
            }

        return issues

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for serialization.

        Returns:
            Dictionary representation
        """
        return {
            "score": self.score,
            "blur_score": self.blur_score,
            "lighting_score": self.lighting_score,
            "face_size": self.face_size,
            "is_acceptable": self.is_acceptable,
            "quality_level": self.get_quality_level(),
            "issues": self.get_issues(),
            "yaw": self.yaw,
            "pitch": self.pitch,
            "pose_acceptable": self.pose_acceptable,
        }

File: demo-ui/src/lib/biometric/domain/entities/quality-assessment.ts
/**
 * Quality assessment value object.
 *
 * Mirrors: app/domain/entities/quality_assessment.py → QualityAssessment
 *
 * Immutable — validated at construction via static factory.
 *
 * Quality Guidelines:
 *   - score 0-40:  Poor (reject)
 *   - score 41-70: Fair (warn user)
 *   - score 71-100: Good (accept)
 */

import type { QualityIssue } from '../types';

export interface QualityParams {
  readonly score: number;
  readonly blurScore: number;
  readonly lightingScore: number;
  readonly faceSize: number;
  readonly isAcceptable: boolean;
}

export type QualityLevel = 'poor' | 'fair' | 'good';

export class QualityAssessment {
  readonly score: number;
  readonly blurScore: number;
  readonly lightingScore: number;
  readonly faceSize: number;
  readonly isAcceptable: boolean;

  private constructor(params: QualityParams) {
    this.score = params.score;
    this.blurScore = params.blurScore;
    this.lightingScore = params.lightingScore;
    this.faceSize = params.faceSize;
    this.isAcceptable = params.isAcceptable;
  }

  /**
   * Create a validated QualityAssessment.
   *
   * @throws {Error} if invariants are violated:
   *   - score must be in [0, 100]
   *   - blurScore must be >= 0
   *   - faceSize must be >= 0
   */
  static create(params: QualityParams): QualityAssessment {
    if (params.score < 0 || params.score > 100) {
      throw new Error(`Score must be between 0 and 100, got ${params.score}`);
    }
    if (params.blurScore < 0) {
      throw new Error(`Blur score cannot be negative: ${params.blurScore}`);
    }
    if (params.faceSize < 0) {
      throw new Error(`Face size cannot be negative: ${params.faceSize}`);
    }
    return new QualityAssessment(params);
  }

  /** Get quality level based on score thresholds. */
  getQualityLevel(): QualityLevel {
    if (this.score < 40) return 'poor';
    if (this.score < 71) return 'fair';
    return 'good';
  }

  /** Check if the image is too blurry. */
  isBlurry(threshold = 100): boolean {
    return this.blurScore < threshold;
  }

  /** Check if the face is too small in the image. */
  isTooSmall(minSize = 80): boolean {
    return this.faceSize < minSize;
  }

  /** Check if lighting conditions are poor. */
  isPoorLighting(threshold = 50): boolean {
    return this.lightingScore < threshold;
  }

  /** Collect all detected quality issues. */
  getIssues(
    blurThreshold = 100,
    minFaceSize = 80,
    lightingThreshold = 50,
  ): QualityIssue[] {
    const issues: QualityIssue[] = [];

    if (this.isBlurry(blurThreshold)) {
      issues.push({
        type: 'blur',
        description: 'Image is too blurry',
        score: this.blurScore,
        threshold: blurThreshold,
      });
    }

    if (this.isTooSmall(minFaceSize)) {
      issues.push({
        type: 'face_size',
        description: 'Face is too small — move closer to camera',
        score: this.faceSize,
        threshold: minFaceSize,
      });
    }

    if (this.isPoorLighting(lightingThreshold)) {
      issues.push({
        type: 'lighting',
        description: 'Poor lighting conditions',
        score: this.lightingScore,
        threshold: lightingThreshold,
      });
    }

    return issues;
  }

  /** Convert to plain object for serialization. */
  toJSON(): QualityParams & { qualityLevel: QualityLevel } {
    return {
      score: this.score,
      blurScore: this.blurScore,
      lightingScore: this.lightingScore,
      faceSize: this.faceSize,
      isAcceptable: this.isAcceptable,
      qualityLevel: this.getQualityLevel(),
    };
  }
}
